lib/src/main/java/com/basistheory/android/service/CardBrandEnricher.kt

```diff
@@ -0,0 +1,215 @@
+                "401179" to null,
+                "438935" to null,
+                "457631" to null,
+                "457632" to null,
+                "431274" to null,
+                "451416" to null,
+                "457393" to null,
+                "504175" to null,
+                "627780" to null,
+                "636297" to null,
+                "636368" to null,
+                "506699" to "506778",
+                "509000" to "509999",
+                "650031" to "650033",
+                "650035" to "650051",
+                "650405" to "650439",
+                "650485" to "650538",
+                "650541" to "650598",
+                "650700" to "650718",
+                "650720" to "650727",
+                "650901" to "650978",
+                "651652" to "651679",
+                "655000" to "655019",
+                "655021" to "655058",
+            ), intArrayOf(16), "###", CardMasks.MASK_4_8_12GAPS_16LENGTH
+        ),
+
+        CardDetails(
+            "mir",
+            listOf("2200" to "2204"),
+            intArrayOf(16, 17, 18, 19),
+            "###",
+            CardMasks.MASK_4_8_12GAPS_19LENGTH
+        ),
+
+        CardDetails(
+            "hiper", listOf(
+                "637095" to null,
+                "63737423" to null,
+                "63743358" to null,
+                "637568" to null,
+                "637599" to null,
+                "637609" to null,
+                "637612" to null,
+            ), intArrayOf(16), "###", CardMasks.MASK_4_8_12GAPS_16LENGTH
+        ),
+
+        CardDetails(
+            "hipercard", listOf(
+                "606282" to null,
+            ), intArrayOf(16), "###", CardMasks.MASK_4_8_12GAPS_16LENGTH
+        )
+    )
+
+    fun evaluateCard(number: String?): CardResult {
+        if(number.isNullOrBlank()) return CardResult(null, false)
+
+        var longestMatch = 0
+        val bestMatch = CardResult(null, false)
+
+        cardBrands.forEach { cardDetails ->
+            cardDetails.identifierRanges.forEach { range ->
+                val possibleMatch = number.take(range.first.length)
+
+                if (!range.second.isNullOrBlank() && range.first.toInt() <= possibleMatch.toInt() && possibleMatch.toInt() <= range.second!!.toInt()) {
+                    longestMatch = considerBestMatchUpdate(
+                        longestMatch,
+                        possibleMatch,
+                        bestMatch,
+                        cardDetails,
+                        number
+                    )
+                } else if (range.first == possibleMatch) {
+                    longestMatch = considerBestMatchUpdate(
+                        longestMatch,
+                        possibleMatch,
+                        bestMatch,
+                        cardDetails,
+                        number
+                    )
+                }
+            }
+        }
+
+        return bestMatch
+    }
```

Comment:
I didn't work out the details, but I have a gut feel that this could be simplified if you add length to the `CardResult` and change `considerBestMatchUpdate` to only take in the current best match + the new candidate match, and return the "best" of the two. Then you can only hold onto one piece of state in this outer loop (`bestMatch`), and you can avoid mutating the bestMatch object in `considerBestMatchUpdate`.

Reply:
The other thing `considerBestMatchUpdate` is doing is computing `complete`, which could actually be a computed property on either `CardResult` or `CardDetails` - no need to compute and mutate this state from outside the class.